pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -111,17 +130,97 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, t
 	return controller.Result{}, nil
 }
 
-// ReconcileControlPlane applies the control plane CAPI objects to the cluster.
-func (r *Reconciler) ReconcileControlPlane(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+// GenerateSpec generates Tinkerbell control plane and workers spec.
+func (r *Reconciler) GenerateSpec(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
 	spec := tinkerbellScope.ClusterSpec
-	log = log.WithValues("phase", "reconcileControlPlane")
-	log.Info("Applying control plane CAPI objects")
+	log = log.WithValues("phase", "GenerateSpec")
+	log.Info("Generating tinkerbell control plane and workers spec")
+
 	cp, err := tinkerbell.ControlPlaneSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
 	if err != nil {
 		return controller.Result{}, err
 	}
+	tinkerbellScope.ControlPlane = cp
+
+	w, err := tinkerbell.WorkersSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating workers spec")
+	}
+	tinkerbellScope.Workers = w
 
-	return clusters.ReconcileControlPlane(ctx, r.client, toClientControlPlane(cp))
+	return controller.Result{}, nil
+}
+
+// DetectOperation detects change type.
+func (r *Reconciler) DetectOperation(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+	log = log.WithValues("phase", "DetectOperation")
+	log.Info("Detecting operation type")
+
+	currentKCP, err := controller.GetKubeadmControlPlane(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+	if currentKCP == nil {
+		tinkerbellScope.ClusterChange = ReconcileNewCluster
+		return controller.Result{}, nil
+	}
+
+	tinkCurrentKCP := tinkerbell.KubeadmControlPlane{KubeadmControlPlane: currentKCP}
+	tinkDesiredKCP := tinkerbell.KubeadmControlPlane{KubeadmControlPlane: tinkerbellScope.ControlPlane.KubeadmControlPlane}
+	wantVersionChange := tinkerbell.HasKubernetesVersionChange(&tinkCurrentKCP, &tinkDesiredKCP)
+	cpWantScaleChange := tinkerbell.ReplicasDiff(&tinkCurrentKCP, &tinkDesiredKCP)
+	workerWantScaleChange := r.workerReplicasDiff(ctx, tinkerbellScope)
+
+	if wantVersionChange {
+		if cpWantScaleChange != 0 || workerWantScaleChange {
+			return controller.Result{}, errors.Errorf("cannot perform scale up or down during k8s version change")
+		}
+		tinkerbellScope.ClusterChange = NeedNewNode
+	} else if cpWantScaleChange != 0 || workerWantScaleChange {
+		tinkerbellScope.ClusterChange = NeedMoreOrLessNode
+	} else {
+		return controller.Result{}, errors.Errorf("cannot detect operation type")
+	}
```

Comment:
This is a little confusing because both functions mention `diff` with 1 returning `int` (the diff) and the other returning `bool`. Given the boolean logic, it makes me think we aren't really looking for a replica diff either. 

Here's a suggestion that might also be a little clearer combined with some other suggestions, and that embraces the [return early philosophy](https://dave.cheney.net/practical-go/presentations/qcon-china.html#_return_early_rather_than_nesting_deeply):

```suggestion
	// Ensure we're performing (1) at least 1 operation and (2) that only 1 kind of
	// operation being requested.
	//
	// TODO: Add detail on _why_.
	
	upgrade := tinkerbell.HasKubernetesVersionChange(&tinkCurrentKCP, &tinkDesiredKCP)
	scaleCp := tinkerbell.ReplicasDiff(&tinkCurrentKCP, &tinkDesiredKCP) != 0
	scaleWkrs := r.workersHaveReplicaDiff(ctx, tinkerbellScope)
	scale := scaleCp || scaleWkrs
	
	if !upgrade && !scale {
		return controller.Result{}, errors.Errorf("cannot detect operation type")
	}
	
	if upgrade && scale {
		return controller.Result{}, errors.Errorf("cannot upgrade kubernetes version and scale the cluster concurrently")
	}
	
	switch {
	case upgrade:
		scope.Operation = K8sVersionUpgradeOperation
	case scale:
		scope.Operation = ScaleOperation
	}
```

Reply:
can't really isolating the data retrieval on worker node diff, since we iterate desired worker node group, see if it exists and any replica changes. then break whenever on a change